File: tests/go/credential_vault_e2e_test.go
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package e2e

import (
	"context"
	"encoding/json"
	"os"
	"testing"
	"time"

	"github.com/alibaba/OpenSandbox/sdks/sandbox/go"
	"github.com/stretchr/testify/require"
)

const credentialVaultDefaultTargetHost = "credential-vault-e2e.opensandbox.test"

var credentialVaultSecrets = map[string]string{
	"bearer-token":           "vault-bearer-token",
	"basic-token":            "dXNlcjpwYXNz",
	"api-key-token":          "vault-api-key-token",
	"client-id":              "vault-client-id",
	"client-secret":          "vault-client-secret",
	"runtime-token":          "vault-runtime-token",
	"runtime-token-replaced": "vault-runtime-token-replaced",
}

func TestCredentialVaultInjectsAllAuthTypes(t *testing.T) {
	targetIP := credentialVaultTargetIP(t)
	ctx, sb := createCredentialVaultSandbox(t)

	state, err := sb.CreateCredentialVault(ctx, opensandbox.CredentialVaultCreateRequest{
		Credentials: credentialVaultCredentials(
			"bearer-token",
			"basic-token",
			"api-key-token",
			"client-id",
			"client-secret",
			"runtime-token",
			"runtime-token-replaced",
		),
		Bindings: []opensandbox.CredentialBinding{
			credentialVaultBinding("bearer", "/bearer", opensandbox.CredentialAuth{
				Type:       opensandbox.CredentialAuthBearer,
				Credential: "bearer-token",
			}),
			credentialVaultBinding("basic", "/basic", opensandbox.CredentialAuth{
				Type:       opensandbox.CredentialAuthBasic,
				Credential: "basic-token",
			}),
			credentialVaultBinding("api-key", "/api-key", opensandbox.CredentialAuth{
				Type:       opensandbox.CredentialAuthAPIKey,
				Name:       "X-Api-Key",
				Credential: "api-key-token",
			}),
			credentialVaultBinding("custom-headers", "/custom-headers", opensandbox.CredentialAuth{
				Type: opensandbox.CredentialAuthCustomHeaders,
				Headers: []opensandbox.CustomHeaderEntry{
					{Name: "X-Client-Id", Credential: "client-id"},
					{Name: "X-Client-Secret", Credential: "client-secret"},
				},
			}),
		},
	})
	require.NoError(t, err)

	statePayload, err := json.Marshal(state)
	require.NoError(t, err)
	for _, secret := range credentialVaultSecrets {
		require.NotContains(t, string(statePayload), secret)
	}

	gotAuthTypes := map[string]bool{}
	for _, binding := range state.Bindings {
		if binding.Auth != nil {
			gotAuthTypes[binding.Auth.Type] = true
		}
	}
	require.Equal(t, map[string]bool{
		"bearer":        true,
		"basic":         true,
		"apiKey":        true,
		"customHeaders": true,
	}, gotAuthTypes)

	for _, path := range []string{"/bearer", "/basic", "/api-key", "/custom-headers"} {
		response := credentialVaultCurlJSON(t, ctx, sb, targetIP, path, true)
		require.Equal(t, true, response["ok"])
		require.Equal(t, path[1:], response["case"])
		require.Empty(t, stringSliceFromJSON(t, response["missingOrInvalid"]))
	}
}

func TestCredentialVaultRuntimeMutationAddsReplacesAndDeletesBinding(t *testing.T) {
	targetIP := credentialVaultTargetIP(t)
	ctx, sb := createCredentialVaultSandbox(t)

	state, err := sb.CreateCredentialVault(ctx, opensandbox.CredentialVaultCreateRequest{})
	require.NoError(t, err)
	require.Equal(t, 1, state.Revision)
	require.Empty(t, state.Credentials)
	require.Empty(t, state.Bindings)

	state, err = sb.PatchCredentialVault(ctx, opensandbox.CredentialVaultPatchRequest{
		ExpectedRevision: intPtr(state.Revision),
		Credentials: &opensandbox.CredentialMutationSet{
			Add: credentialVaultCredentials("runtime-token"),
		},
		Bindings: &opensandbox.CredentialBindingMutationSet{
			Add: []opensandbox.CredentialBinding{
				credentialVaultBinding("runtime-added", "/runtime-added", opensandbox.CredentialAuth{
					Type:       opensandbox.CredentialAuthAPIKey,
					Name:       "X-Runtime-Token",
					Credential: "runtime-token",
				}),
			},
		},
	})
	require.NoError(t, err)
	require.Equal(t, 2, state.Revision)
	require.Len(t, state.Credentials, 1)
	require.Equal(t, "runtime-token", state.Credentials[0].Name)
	require.Len(t, state.Bindings, 1)
	require.Equal(t, "runtime-added", state.Bindings[0].Name)

	response := credentialVaultCurlJSON(t, ctx, sb, targetIP, "/runtime-added", true)
	require.Equal(t, true, response["ok"])
	require.Equal(t, "runtime-added", response["case"])
	require.Empty(t, stringSliceFromJSON(t, response["missingOrInvalid"]))

	state, err = sb.PatchCredentialVault(ctx, opensandbox.CredentialVaultPatchRequest{
		ExpectedRevision: intPtr(state.Revision),
		Bindings:         &opensandbox.CredentialBindingMutationSet{Delete: []string{"runtime-added"}},
	})
	require.NoError(t, err)
	require.Equal(t, 3, state.Revision)
	require.Empty(t, state.Bindings)

	state, err = sb.PatchCredentialVault(ctx, opensandbox.CredentialVaultPatchRequest{
		ExpectedRevision: intPtr(state.Revision),
		Credentials: &opensandbox.CredentialMutationSet{
			Replace: []opensandbox.Credential{
				credentialVaultCredential("runtime-token", "runtime-token-replaced"),
			},
		},
		Bindings: &opensandbox.CredentialBindingMutationSet{
			Add: []opensandbox.CredentialBinding{
				credentialVaultBinding("runtime-replaced", "/runtime-replaced", opensandbox.CredentialAuth{
					Type:       opensandbox.CredentialAuthAPIKey,
					Name:       "X-Runtime-Token",
					Credential: "runtime-token",
				}),
			},
		},
	})
	require.NoError(t, err)
	require.Equal(t, 4, state.Revision)
	require.Len(t, state.Credentials, 1)
	require.Equal(t, "runtime-token", state.Credentials[0].Name)
	require.Len(t, state.Bindings, 1)
	require.Equal(t, "runtime-replaced", state.Bindings[0].Name)

	statePayload, err := json.Marshal(state)
	require.NoError(t, err)
	require.NotContains(t, string(statePayload), credentialVaultSecrets["runtime-token"])
	require.NotContains(t, string(statePayload), credentialVaultSecrets["runtime-token-replaced"])

	response = credentialVaultCurlJSON(t, ctx, sb, targetIP, "/runtime-replaced", true)
	require.Equal(t, true, response["ok"])
	require.Equal(t, "runtime-replaced", response["case"])
	require.Empty(t, stringSliceFromJSON(t, response["missingOrInvalid"]))

	response = credentialVaultCurlJSON(t, ctx, sb, targetIP, "/runtime-added", false)
	require.Equal(t, false, response["ok"])
	require.Equal(t, "runtime-added", response["case"])
	require.Equal(t, []string{"x-runtime-token"}, stringSliceFromJSON(t, response["missingOrInvalid"]))

	state, err = sb.PatchCredentialVault(ctx, opensandbox.CredentialVaultPatchRequest{
		ExpectedRevision: intPtr(state.Revision),
		Bindings:         &opensandbox.CredentialBindingMutationSet{Delete: []string{"runtime-replaced"}},
	})
	require.NoError(t, err)
	require.Equal(t, 5, state.Revision)
	require.Empty(t, state.Bindings)

	state, err = sb.PatchCredentialVault(ctx, opensandbox.CredentialVaultPatchRequest{
		ExpectedRevision: intPtr(state.Revision),
		Credentials:      &opensandbox.CredentialMutationSet{Delete: []string{"runtime-token"}},
	})
	require.NoError(t, err)
	require.Equal(t, 6, state.Revision)
	require.Empty(t, state.Credentials)
}

func credentialVaultTargetHost() string {
	if host := os.Getenv("OPENSANDBOX_CREDENTIAL_VAULT_E2E_TARGET_HOST"); host != "" {
		return host
	}
	return credentialVaultDefaultTargetHost
}

func credentialVaultTargetIP(t *testing.T) string {
	t.Helper()
	targetIP := os.Getenv("OPENSANDBOX_CREDENTIAL_VAULT_E2E_TARGET_IP")
	if targetIP == "" {
		t.Skip("set OPENSANDBOX_CREDENTIAL_VAULT_E2E_TARGET_IP to run Credential Vault E2E")
	}
	return targetIP
}

func createCredentialVaultSandbox(t *testing.T) (context.Context, *opensandbox.Sandbox) {
	t.Helper()

	image := os.Getenv("OPENSANDBOX_CREDENTIAL_VAULT_E2E_SANDBOX_IMAGE")
	if image == "" {
		image = getSandboxImage()
	}

	config := connectionConfigForStreaming(t)
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	t.Cleanup(cancel)

	sb, err := opensandbox.CreateSandbox(ctx, config, opensandbox.SandboxCreateOptions{
		Image:          image,
		ResourceLimits: credentialVaultSandboxResource(),
		ReadyTimeout:   90 * time.Second,
		NetworkPolicy: &opensandbox.NetworkPolicy{
			DefaultAction: "allow",
			Egress:        []opensandbox.NetworkRule{{Action: "allow", Target: credentialVaultTargetHost()}},
		},
		CredentialProxy: &opensandbox.CredentialProxyConfig{Enabled: true},
		Metadata: map[string]string{
			credentialVaultLabelKey(): credentialVaultLabelValue(),
		},
	})
	require.NoError(t, err)
	t.Cleanup(func() { _ = sb.Kill(context.Background()) })
	return ctx, sb
}

func credentialVaultSandboxResource() opensandbox.ResourceLimits {
	cpu := os.Getenv("OPENSANDBOX_E2E_SANDBOX_CPU")
	if cpu == "" {
		cpu = "1"
	}
	memory := os.Getenv("OPENSANDBOX_E2E_SANDBOX_MEMORY")
	if memory == "" {
		memory = "2Gi"
	}
	return opensandbox.ResourceLimits{"cpu": cpu, "memory": memory}
}

func credentialVaultCredentials(names ...string) []opensandbox.Credential {
	credentials := make([]opensandbox.Credential, 0, len(names))
	for _, name := range names {
		credentials = append(credentials, credentialVaultCredential(name, name))
	}
	return credentials
}

func credentialVaultCredential(name, valueName string) opensandbox.Credential {
	return opensandbox.Credential{
		Name: name,
		Source: opensandbox.InlineCredentialSource{
			Type:  opensandbox.CredentialSourceInline,
			Value: credentialVaultSecrets[valueName],
		},
	}
}

func credentialVaultBinding(name, path string, auth opensandbox.CredentialAuth) opensandbox.CredentialBinding {
	return opensandbox.CredentialBinding{
		Name: name,
		Match: opensandbox.CredentialMatch{
			Schemes: []opensandbox.CredentialScheme{opensandbox.CredentialSchemeHTTP},
			Ports:   []int{80},
			Hosts:   []string{credentialVaultTargetHost()},
			Methods: []string{"GET"},
			Paths:   []string{path},
		},
		Auth: auth,
	}
}

func credentialVaultCurlJSON(
	t *testing.T,
	ctx context.Context,
	sb *opensandbox.Sandbox,
	targetIP string,
	path string,
	failOnHTTPError bool,
) map[string]any {
	t.Helper()
	failFlag := ""
	if failOnHTTPError {
		failFlag = "--fail "
	}
	command := "curl " + failFlag +
		"--silent --show-error --connect-timeout 5 --max-time 20 " +
		"--resolve " + credentialVaultTargetHost() + ":80:" + targetIP + " " +
		"http://" + credentialVaultTargetHost() + path
	for _, secret := range credentialVaultSecrets {
		require.NotContains(t, command, secret)
	}

	exec, err := sb.RunCommand(ctx, command, nil)
	require.NoError(t, err)
	require.Nil(t, exec.Error)
	require.NotNil(t, exec.ExitCode)
	require.Equal(t, 0, *exec.ExitCode)

	stdout := exec.Text()
	require.NotEmpty(t, stdout)

	var response map[string]any
	require.NoError(t, json.Unmarshal([]byte(stdout), &response))
	return response
}

func stringSliceFromJSON(t *testing.T, value any) []string {
	t.Helper()
	values, ok := value.([]any)
	require.Truef(t, ok, "expected []any, got %T", value)
	result := make([]string, 0, len(values))
	for _, value := range values {
		text, ok := value.(string)
		require.Truef(t, ok, "expected string, got %T", value)
		result = append(result, text)
	}
	return result
}

func credentialVaultLabelKey() string {
	if key := os.Getenv("OPENSANDBOX_CREDENTIAL_VAULT_E2E_LABEL_KEY"); key != "" {
		return key
	}
	return "opensandbox.e2e"
}

func credentialVaultLabelValue() string {
	if value := os.Getenv("OPENSANDBOX_CREDENTIAL_VAULT_E2E_LABEL_VALUE"); value != "" {
		return value
	}
	return "credential-vault"
}

func intPtr(v int) *int {
	return &v
}

File: tests/java/src/test/java/com/alibaba/opensandbox/e2e/CredentialVaultE2ETest.java
/*
 * Copyright 2026 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.e2e;

import static org.junit.jupiter.api.Assertions.*;

import com.alibaba.opensandbox.sandbox.Sandbox;
import com.alibaba.opensandbox.sandbox.domain.models.execd.executions.Execution;
import com.alibaba.opensandbox.sandbox.domain.models.execd.executions.OutputMessage;
import com.alibaba.opensandbox.sandbox.domain.models.execd.executions.RunCommandRequest;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.Credential;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialAuth;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialBinding;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialBindingMetadata;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialBindingMutationSet;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialMatch;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialMetadata;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialMutationSet;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialProxyConfig;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialVaultCreateRequest;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialVaultPatchRequest;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CredentialVaultState;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.CustomHeaderEntry;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.NetworkPolicy;
import com.alibaba.opensandbox.sandbox.domain.models.sandboxes.NetworkRule;
import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

@Tag("e2e")
@DisplayName("Credential Vault E2E Tests (JVM/Kotlin SDK)")
class CredentialVaultE2ETest extends BaseE2ETest {

    private static final String DEFAULT_TARGET_HOST = "credential-vault-e2e.opensandbox.test";
    private static final Map<String, String> SECRET_VALUES =
            Map.of(
                    "bearer-token", "vault-bearer-token",
                    "basic-token", "dXNlcjpwYXNz",
                    "api-key-token", "vault-api-key-token",
                    "client-id", "vault-client-id",
                    "client-secret", "vault-client-secret",
                    "runtime-token", "vault-runtime-token",
                    "runtime-token-replaced", "vault-runtime-token-replaced");

    @Test
    @Timeout(value = 5, unit = TimeUnit.MINUTES)
    void credentialVaultInjectsAllAuthTypes() {
        String targetIp = credentialVaultTargetIp();
        Sandbox sandbox = createCredentialVaultSandbox();

        try {
            CredentialVaultState state =
                    sandbox.credentialVault()
                            .create(
                                    CredentialVaultCreateRequest.builder()
                                            .credentials(
                                                    credentials(
                                                            "bearer-token",
                                                            "basic-token",
                                                            "api-key-token",
                                                            "client-id",
                                                            "client-secret",
                                                            "runtime-token",
                                                            "runtime-token-replaced"))
                                            .bindings(
                                                    List.of(
                                                            binding(
                                                                    "bearer",
                                                                    "/bearer",
                                                                    CredentialAuth.bearer(
                                                                            "bearer-token")),
                                                            binding(
                                                                    "basic",
                                                                    "/basic",
                                                                    CredentialAuth.basic(
                                                                            "basic-token")),
                                                            binding(
                                                                    "api-key",
                                                                    "/api-key",
                                                                    CredentialAuth.apiKey(
                                                                            "X-Api-Key",
                                                                            "api-key-token")),
                                                            binding(
                                                                    "custom-headers",
                                                                    "/custom-headers",
                                                                    CredentialAuth.customHeaders(
                                                                            List.of(
                                                                                    CustomHeaderEntry
                                                                                            .builder()
                                                                                            .name(
                                                                                                    "X-Client-Id")
                                                                                            .credential(
                                                                                                    "client-id")
                                                                                            .build(),
                                                                                    CustomHeaderEntry
                                                                                            .builder()
                                                                                            .name(
                                                                                                    "X-Client-Secret")
                                                                                            .credential(
                                                                                                    "client-secret")
                                                                                            .build())))))
                                            .build());

            Set<String> authTypes =
                    state.getBindings().stream()
                            .map(CredentialBindingMetadata::getAuth)
                            .filter(auth -> auth != null)
                            .map(auth -> auth.getType())
                            .collect(Collectors.toSet());
            assertEquals(Set.of("bearer", "basic", "apiKey", "customHeaders"), authTypes);

            for (String path : List.of("/bearer", "/basic", "/api-key", "/custom-headers")) {
                String response = curlJson(sandbox, targetIp, path, true);
                assertJsonCase(response, path.substring(1), true, "[]");
            }
        } finally {
            killSandbox(sandbox);
        }
    }

    @Test
    @Timeout(value = 5, unit = TimeUnit.MINUTES)
    void credentialVaultRuntimeMutationAddsReplacesAndDeletesBinding() {
        String targetIp = credentialVaultTargetIp();
        Sandbox sandbox = createCredentialVaultSandbox();

        try {
            CredentialVaultState state =
                    sandbox.credentialVault()
                            .create(
                                    CredentialVaultCreateRequest.builder()
                                            .credentials(List.of())
                                            .bindings(List.of())
                                            .build());
            assertEquals(1, state.getRevision());
            assertTrue(state.getCredentials().isEmpty());
            assertTrue(state.getBindings().isEmpty());

            state =
                    sandbox.credentialVault()
                            .patch(
                                    CredentialVaultPatchRequest.builder()
                                            .expectedRevision(state.getRevision())
                                            .credentials(
                                                    CredentialMutationSet.builder()
                                                            .add(
                                                                    List.of(
                                                                            credential(
                                                                                    "runtime-token",
                                                                                    "runtime-token")))
                                                            .build())
                                            .bindings(
                                                    CredentialBindingMutationSet.builder()
                                                            .add(
                                                                    List.of(
                                                                            binding(
                                                                                    "runtime-added",
                                                                                    "/runtime-added",
                                                                                    CredentialAuth
                                                                                            .apiKey(
                                                                                                    "X-Runtime-Token",
                                                                                                    "runtime-token"))))
                                                            .build())
                                            .build());
            assertEquals(2, state.getRevision());
            assertEquals(List.of("runtime-token"), credentialNames(state.getCredentials()));
            assertEquals(List.of("runtime-added"), bindingNames(state.getBindings()));

            String response = curlJson(sandbox, targetIp, "/runtime-added", true);
            assertJsonCase(response, "runtime-added", true, "[]");

            state =
                    sandbox.credentialVault()
                            .patch(
                                    CredentialVaultPatchRequest.builder()
                                            .expectedRevision(state.getRevision())
                                            .bindings(
                                                    CredentialBindingMutationSet.builder()
                                                            .delete("runtime-added")
                                                            .build())
                                            .build());
            assertEquals(3, state.getRevision());
            assertTrue(state.getBindings().isEmpty());

            state =
                    sandbox.credentialVault()
                            .patch(
                                    CredentialVaultPatchRequest.builder()
                                            .expectedRevision(state.getRevision())
                                            .credentials(
                                                    CredentialMutationSet.builder()
                                                            .replace(
                                                                    List.of(
                                                                            credential(
                                                                                    "runtime-token",
                                                                                    "runtime-token-replaced")))
                                                            .build())
                                            .bindings(
                                                    CredentialBindingMutationSet.builder()
                                                            .add(
                                                                    List.of(
                                                                            binding(
                                                                                    "runtime-replaced",
                                                                                    "/runtime-replaced",
                                                                                    CredentialAuth
                                                                                            .apiKey(
                                                                                                    "X-Runtime-Token",
                                                                                                    "runtime-token"))))
                                                            .build())
                                            .build());
            assertEquals(4, state.getRevision());
            assertEquals(List.of("runtime-token"), credentialNames(state.getCredentials()));
            assertEquals(List.of("runtime-replaced"), bindingNames(state.getBindings()));

            response = curlJson(sandbox, targetIp, "/runtime-replaced", true);
            assertJsonCase(response, "runtime-replaced", true, "[]");

            response = curlJson(sandbox, targetIp, "/runtime-added", false);
            assertJsonCase(response, "runtime-added", false, "[\"x-runtime-token\"]");

            state =
                    sandbox.credentialVault()
                            .patch(
                                    CredentialVaultPatchRequest.builder()
                                            .expectedRevision(state.getRevision())
                                            .bindings(
                                                    CredentialBindingMutationSet.builder()
                                                            .delete("runtime-replaced")
                                                            .build())
                                            .build());
            assertEquals(5, state.getRevision());
            assertTrue(state.getBindings().isEmpty());

            state =
                    sandbox.credentialVault()
                            .patch(
                                    CredentialVaultPatchRequest.builder()
                                            .expectedRevision(state.getRevision())
                                            .credentials(
                                                    CredentialMutationSet.builder()
                                                            .delete("runtime-token")
                                                            .build())
                                            .build());
            assertEquals(6, state.getRevision());
            assertTrue(state.getCredentials().isEmpty());
        } finally {
            killSandbox(sandbox);
        }
    }

    private static Sandbox createCredentialVaultSandbox() {
        String image =
                envOrDefault(
                        "OPENSANDBOX_CREDENTIAL_VAULT_E2E_SANDBOX_IMAGE", getSandboxImage());
        Map<String, String> resource = new HashMap<>();
        resource.put("cpu", envOrDefault("OPENSANDBOX_E2E_SANDBOX_CPU", "1"));
        resource.put("memory", envOrDefault("OPENSANDBOX_E2E_SANDBOX_MEMORY", "2Gi"));

        return Sandbox.builder()
                .connectionConfig(createConnectionConfig(false))
                .image(image)
                .resource(resource)
                .timeout(Duration.ofMinutes(5))
                .readyTimeout(Duration.ofSeconds(90))
                .networkPolicy(
                        NetworkPolicy.builder()
                                .defaultAction(NetworkPolicy.DefaultAction.ALLOW)
                                .addEgress(
                                        NetworkRule.builder()
                                                .action(NetworkRule.Action.ALLOW)
                                                .target(credentialVaultTargetHost())
                                                .build())
                                .build())
                .credentialProxy(CredentialProxyConfig.enabled())
                .metadata(
                        Map.of(
                                envOrDefault(
                                        "OPENSANDBOX_CREDENTIAL_VAULT_E2E_LABEL_KEY",
                                        "opensandbox.e2e"),
                                envOrDefault(
                                        "OPENSANDBOX_CREDENTIAL_VAULT_E2E_LABEL_VALUE",
                                        "credential-vault")))
                .build();
    }

    private static List<Credential> credentials(String... names) {
        return java.util.Arrays.stream(names)
                .map(name -> credential(name, name))
                .collect(Collectors.toList());
    }

    private static Credential credential(String name, String valueName) {
        return Credential.builder().name(name).inlineSource(SECRET_VALUES.get(valueName)).build();
    }

    private static CredentialBinding binding(String name, String path, CredentialAuth auth) {
        return CredentialBinding.builder()
                .name(name)
                .match(
                        CredentialMatch.builder()
                                .schemes(CredentialMatch.Scheme.HTTP)
                                .ports(80)
                                .hosts(credentialVaultTargetHost())
                                .methods("GET")
                                .paths(path)
                                .build())
                .auth(auth)
                .build();
    }

    private static String curlJson(
            Sandbox sandbox, String targetIp, String path, boolean failOnHttpError) {
        String failFlag = failOnHttpError ? "--fail " : "";
        String command =
                "curl "
                        + failFlag
                        + "--silent --show-error --connect-timeout 5 --max-time 20 "
                        + "--resolve "
                        + credentialVaultTargetHost()
                        + ":80:"
                        + targetIp
                        + " http://"
                        + credentialVaultTargetHost()
                        + path;
        for (String secret : SECRET_VALUES.values()) {
            assertFalse(command.contains(secret), "command must not contain secret material");
        }

        Execution execution =
                sandbox.commands().run(RunCommandRequest.builder().command(command).build());
        assertNull(execution.getError(), "curl command failed");
        assertEquals(0, execution.getExitCode());
        String stdout =
                execution.getLogs().getStdout().stream()
                        .map(OutputMessage::getText)
                        .collect(Collectors.joining());
        assertFalse(stdout.isBlank(), "curl response must not be blank");
        return stdout;
    }

    private static void assertJsonCase(
            String payload, String expectedCase, boolean expectedOk, String expectedMissing) {
        assertTrue(payload.contains("\"ok\":" + expectedOk), payload);
        assertTrue(payload.contains("\"case\":\"" + expectedCase + "\""), payload);
        assertTrue(payload.contains("\"missingOrInvalid\":" + expectedMissing), payload);
    }

    private static List<String> credentialNames(List<CredentialMetadata> credentials) {
        return credentials.stream().map(CredentialMetadata::getName).collect(Collectors.toList());
    }

    private static List<String> bindingNames(List<CredentialBindingMetadata> bindings) {
        return bindings.stream().map(CredentialBindingMetadata::getName).collect(Collectors.toList());
    }

    private static String credentialVaultTargetHost() {
        return envOrDefault("OPENSANDBOX_CREDENTIAL_VAULT_E2E_TARGET_HOST", DEFAULT_TARGET_HOST);
    }

    private static String credentialVaultTargetIp() {
        String targetIp = System.getenv("OPENSANDBOX_CREDENTIAL_VAULT_E2E_TARGET_IP");
        Assumptions.assumeTrue(
                targetIp != null && !targetIp.isBlank(),
                "Set OPENSANDBOX_CREDENTIAL_VAULT_E2E_TARGET_IP to run Credential Vault E2E");
        return targetIp;
    }

    private static String envOrDefault(String name, String defaultValue) {
        String value = System.getenv(name);
        return value == null || value.isBlank() ? defaultValue : value;
    }

    private static void killSandbox(Sandbox sandbox) {
        try {
            sandbox.kill();
        } catch (Exception ignored) {
        }
        try {
            sandbox.close();
        } catch (Exception ignored) {
        }
    }
}
